### PR TITLE
feat: Add Golang support for EventBridge Schemas templates

### PR DESCRIPTION
**Pull request by kolaamoo25**
*Issue #, if available:* N/A

*Description of changes:* Adding Golang templates for EventBridge Schemas.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/integration/build_invoke/test_build_invoke_go1_x.py [hunk 1]
```diff
@@ -15,6 +16,14 @@ class BuildInvoke_go1_x_cookiecutter_aws_sam_hello_golang(BuildInvokeBase.BuildI
     use_container = False
     directory = "go1.x/cookiecutter-aws-sam-hello-golang"
 
+class BuildInvoke_go1_x_cookiecutter_aws_sam_eventbridge_hello_golang(BuildInvokeBase.BuildInvokeBase):
+    use_container = False
+    directory = "go1.x/cookiecutter-aws-sam-eventbridge-hello-golang"
+
+@skip("eventbridge schema app requires credential to pull missing files, skip")
```
Comment (moelasmar) [inline tests/integration/build_invoke/test_build_invoke_go1_x.py:23]: could you please explain why this sample requires credential. I could not see any logic in `main.go` that it requires credential.

**Review by moelasmar — CHANGES_REQUESTED** [review]

File: tests/integration/build_invoke/test_build_invoke_go1_x.py (hunk 1 above)
Comment (moelasmar) [reply tests/integration/build_invoke/test_build_invoke_go1_x.py:23]: I got it, the retrieval step done during `sam init` as it is a dynamic template.

Comment (moelasmar) [reply tests/integration/build_invoke/test_build_invoke_go1_x.py:23]: is it ok to remove this test case, it will not be used anyway, and add it to sam cli instead.

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+```bash
+go test -v ./hello-world/
+```
+
+## Cleanup
+
+To delete the sample application that you created, use the AWS CLI. Assuming you used your project name for the stack name, you can run the following:
+
+```bash
+aws cloudformation delete-stack --stack-name {{ cookiecutter.project_name }}
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md:115]: Will update with sam delete, for all eventbridge app.

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/hello-world/main.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-lambda-go/lambda"
+	"{{ cookiecutter.project_name }}/hello-world/model/aws/ec2"
+	"{{ cookiecutter.project_name }}/hello-world/model/aws/ec2/marshaller"
+)
+
+func handler(context context.Context, awsEvent ec2.AWSEvent) ([]byte , error) {
+	/*
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/hello-world/main.go:12]: This is my first time making a contribution to this repo. Where would these docs be hosted and could you give me an example of one?

File: manifest.json [hunk 4]
```diff
@@ -74,6 +74,23 @@
...
+      "directory": "go1.x/cookiecutter-aws-sam-eventbridge-hello-golang",
+      "displayName": "EventBridge Hello World",
+      "dependencyManager": "mod",
+      "appTemplate": "eventBridge-hello-world",
+      "packageType": "Zip",
+      "useCaseName": "Infrastructure event management"
+    },
+    {
+      "directory": "go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang",
+      "displayName": "EventBridge App from scratch (100+ Event Schemas)",
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply manifest.json:87]: When customers use the sam init workflow to create EventBridge applications they have the option of selecting a registry to pull the schema from. EventBridge Schema Registry supports a first party registry which holds schemas from AWS services. In that case we currently have over 100 schemas on-boarded to the service and available for all customers to use. We currently use that display name for all of the other custom app templates (for Java and Python versions).

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/hello-world/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (moelasmar) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/hello-world/main.go:12]: I mean a comment to the function itself

File: tests/integration/build_invoke/test_build_invoke_go1_x.py (hunk 1 above)
Comment (moelasmar) [reply tests/integration/build_invoke/test_build_invoke_go1_x.py:23]: Hmmm .. it is ok, but can you add todo comment to remove the skip once we start providing

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (moelasmar) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md:115]: Thanks

File: manifest.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (moelasmar) [reply manifest.json:87]: Got it. Thanks

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# Cookiecutter SAM for golang Lambda functions
+
+This is a [Cookiecutter](https://github.com/audreyr/cookiecutter) template to create a Serverless EventBridge Hello World App based on Serverless Application Model (SAM) and Golang.
+
+It is important to note that you should not try to `git clone` this project but use `cookiecutter` CLI instead as ``{{cookiecutter.project_name}}`` will be rendered based on your input and therefore all variables and files will be rendered properly.
+
+This template creates a Serverless Application that reacts to EC2 Instance State change events, demonstrating the power of event-driven development with Amazon EventBridge.
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/README.md:7]: I can change up the README more to differentiate that this is a dynamic template while the other one is a static template.

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/tests/test_bake_project.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+def test_app_content(cookies):
+    result = cookies.bake(extra_context={'project_name': 'test_project'})
+    app_file = result.project.join('hello-world', 'main.go')
+    app_content = app_file.readlines()
+    app_content = ''.join(app_content)
+
+    contents = (
+        "github.com/aws/aws-lambda-go/events",
+        "resp, err := http.Get(DefaultHTTPGetAddress)",
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/tests/test_bake_project.py:43]: This was just a boilerplate test that was included in the standard golang hello world example. I can remove it from here as it is not directly needed.

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+Transform: AWS::Serverless-2016-10-31
+Description: >
+  {{ cookiecutter.project_name }}
+  Sample SAM Template for {{ cookiecutter.project_name }}
+# More info about Globals: https://github.com/awslabs/serverless-application-model/blob/master/docs/globals.rst
+Globals:
+  Function:
+    Timeout: 20
+
+Resources:
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml:11]: Since this example is a dynamic template we don't know the resource the customer is interacting with ahead of time.

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/{{cookiecutter.function_name}}/main.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-lambda-go/lambda"
+	"{{ cookiecutter.codegen_path }}"
+)
+
+func handler(context context.Context, event {{ cookiecutter.codegen_package_name }}.AWSEvent) ([]byte , error) {
+	/*
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/{{cookiecutter.function_name}}/main.go:11]: Noted, will remove and update with documentation comments.

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (moelasmar) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml:11]: My issue is this sample example is not testable, the customer needs to do prerequisite steps to test the end to end scenario.

[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml:11]: Is the purpose of these tests to be able to test end to end workflow? From the examples that I created these templates from in the past we have not created the AWS Resource the original event originated from.

[the remark this answers is not in the document]
Comment (moelasmar) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml:11]: These samples should let the customer knows how to build a specific use case. 
So, in my opinion as a customer I would like to have a complete running basic sample, it shows me how every thing works together.

[the remark this answers is not in the document]
Comment (moelasmar) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml:11]: @praneetap .. what do you think?

[the remark this answers is not in the document]
Comment (moelasmar) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml:11]: Ok, I checked this internally. it is Ok to have this sample as it is as long as we include the steps to test this sample end to end in the read me file.

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md]: All EventBridge application README files have been updated to have sam delete instead

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/{{cookiecutter.project_name}}/template.yaml:11]: I've updated the README of the hello world app with instructions on how to end to end test the integration of EC2 instance change notifications and the lambda function that is deployed

File: tests/integration/build_invoke/test_build_invoke_go1_x.py (hunk 1 above)
Comment (kolaamoo25) [reply tests/integration/build_invoke/test_build_invoke_go1_x.py:24]: Todo added to unskip tests when they are run in an environment with AWS Credentials

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/tests/test_bake_project.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+def test_app_test_content(cookies):
+    result = cookies.bake(extra_context={'project_name': 'test_project'})
+    app_file = result.project.join('hello-world', 'main_test.go')
+    app_content = app_file.readlines()
+    app_content = ''.join(app_content)
+
+    contents = (
+        "DefaultHTTPGetAddress = \"http://127.0.0.1:12345\"",
+        "DefaultHTTPGetAddress = ts.URL",
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/tests/test_bake_project.py]: Removed files that were not used.

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/template.yaml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+Transform: AWS::Serverless-2016-10-31
+Description: >
+  {{ cookiecutter.project_name }}
+  Sample SAM Template for {{ cookiecutter.project_name }}
+# More info about Globals: https://github.com/awslabs/serverless-application-model/blob/master/docs/globals.rst
+Globals:
+  Function:
+    Timeout: 20
+
+Resources:
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/template.yaml:11]: Updated README file of static template with how to E2E the integration between EC2 Instance Change notifications and the lambda function that is deployed.

**Review by moelasmar — APPROVED** [review]

File: manifest.json [hunk 11]
```diff
@@ -74,6 +74,23 @@
       "packageType": "Zip",
       "useCaseName": "Hello World Example"
     },
+    {
```
Comment (super132) [inline manifest.json:77]: This file should not be modified. For new templates, please modify `manifest_v2.json` instead.

Comment (kolaamoo25) [reply manifest.json]: Moved template info to new manifest_v2.json file

**Review by jfuss — CHANGES_REQUESTED** [review]
Some comments. For the future, please do not touch every example for updates if you are adding a feature. It makes it really hard to follow and if this gets reverted, we revert those changes as well. I won't make you revert them but be aware for the future.

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+## Unit tests
+
+We use `testing` package that is built-in in Golang and you can simply run the following commands to run the tests:
+
+```bash
+go test -v ./hello-world/
+```
+
+## Steps to End to End (E2E) test using the AWS Console
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md:110]: Will keep that in mind for all future commits. Thanks

[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md:110]: I can remove the steps to E2E test in the console. The addition of that was a suggestion from @moelasmar from below

> Ok, I checked this internally. it is Ok to have this sample as it is as long as we include the steps to test this sample end to end in the read me file.

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/cookiecutter.json [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    "architectures": {
+        "value": []
+    },
+    "function_name": "HelloWorld",
+    "AWS_Schema_registry": "aws.events",
+    "AWS_Schema_name": "EC2InstanceStateChangeNotification",
+    "AWS_Schema_root": "aws.ec2.EC2InstanceStateChangeNotification",
+    "AWS_Schema_source": "aws.ec2",
+    "AWS_Schema_detail_type": "EC2 Instance State-change Notification",
+    "codegen_path": "{{ cookiecutter.project_name }}/{{ cookiecutter.function_name }}/{{ cookiecutter.AWS_Schema_root|lower|replace('.', '/') }}",
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/cookiecutter.json:13]: The way that Golang does imports requires you to know the whole path from the GoRoot, while Python and Java have friendlier import semantics. This allows cookiecutter to pass in the right path to make things build right out of the box

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jfuss) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md:110]: I see. If we really think this is needed, we need to provide ways for customers to remove it. I prefer to have everything in the stack, as there shouldn't be a need for a customers to use `sam init` and then have to go to the console for things (not a great experience). Maybe EC2 is just a bad example and could use something else that would work with SAM Accelerate instead.

If it's a bigger change, I would be more inclined to remove this and find a way to integrate SAM Accelerate and expanding the example.

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/cookiecutter.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (jfuss) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/cookiecutter.json:13]: Interesting. Deeper question for why you need GoRoot? I thought `Modules` allowed you to not think about it at all but maybe there is something deeper here. I am not Go Expert, so more just curious.

File: java11/cookiecutter-aws-sam-eventbridge-hello-java-gradle/{{cookiecutter.project_name}}/README.md [hunk 14]
```diff
@@ -120,10 +120,10 @@ HelloWorldFunction$ gradle test
 
 ## Cleanup
 
-To delete the sample application and the bucket that you created, use the AWS CLI.
+To delete the sample application and the bucket that you created, use the SAM CLI and AWS CLI.
 
 ```bash
-{{ cookiecutter.project_name }}$ aws cloudformation delete-stack --stack-name {{ cookiecutter.project_name }}
+{{ cookiecutter.project_name }}$ sam delete --stack-name {{ cookiecutter.project_name }}
 {{ cookiecutter.project_name }}$ aws s3 rb s3://BUCKET_NAME
```
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply java11/cookiecutter-aws-sam-eventbridge-hello-java-gradle/{{cookiecutter.project_name}}/README.md:127]: Screen Shot 2022-03-07 at 1 07 09 PM

I believe the s3 bucket deletion is part of an added step in the README file. Posted a snapshot above. A bucket might have been made that is not part of the stack deployment using the template.yaml file.

File: go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-hello-golang/{{cookiecutter.project_name}}/README.md:110]: I agree that the customer experience isn't the best having to manually End to End test using the console. We can remove it for now, and I'll create a SIM item for my team to look into a better way of do it (possibly using SAM Accelerate). Then we can expand it to all of the EventBridge app templates instead of just this one.

File: go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/cookiecutter.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (kolaamoo25) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/cookiecutter.json:13]: Sorry for the confusion here, this is not related to GoRoot. This is the path for the schema generated code from where the go module is specified (in this case {{ cookiecutter.project_name }}). To import the generated code into the main.go file you need to know the import path for the generated code. Since this is dynamic based on the schema that is downloaded, this is some post-processing to generate the path at runtime.

[the remark this answers is not in the document]
Comment (jfuss) [reply go1.x/cookiecutter-aws-sam-eventbridge-schema-app-golang/cookiecutter.json:13]: Ahh. Ok. Thanks for explaining it, this makes more sense to me now.

File: java11/cookiecutter-aws-sam-eventbridge-hello-java-gradle/{{cookiecutter.project_name}}/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (jfuss) [reply java11/cookiecutter-aws-sam-eventbridge-hello-java-gradle/{{cookiecutter.project_name}}/README.md:127]: Fair. SAM CLI can manage buckets but updating that is out of scope here. :)

**Review by jfuss — APPROVED** [review]

**Comment by torresxb1** [comment]
Note: this PR shouldn't be merged until the corresponding SAM CLI one is released: https://github.com/aws/aws-sam-cli/pull/3657
